projects/batfish/src/main/java/org/batfish/job/BatfishJobExecutor.java

```diff
@@ -41,105 +50,143 @@ public BatfishJobExecutor(
-          } else {
-            String failureMessage =
-                "Failure running job after elapsed time: "
-                    + time
-                    + "\n-----BEGIN JOB LOG-----\n"
-                    + result
-                        .getHistory()
-                        .toString(BatfishLogger.getLogLevel(_settings.getLogLevel()))
-                    + "\n-----END JOB LOG-----";
-            BatfishException bfc = new BatfishException(failureMessage, failureCause);
-            if (_settings.getExitOnFirstError()) {
-              result.appendHistory(_logger);
-              throw bfc;
-            } else {
-              processingError = true;
-              result.appendHistory(_logger);
-              _logger.error(failureMessage + ":\n\t" + ExceptionUtils.getStackTrace(failureCause));
-              failureCauses.add(bfc);
-              if (!_haltOnProcessingError) {
-                result.applyTo(output, _logger, answerElement);
-              }
-            }
-          }
-        } else {
-          continue;
-        }
-      }
-      if (!futures.isEmpty()) {
+
+    for (int i = 0; i < jobs.size(); i++) {
+      try {
+        // getting the next completed task
+        Future<JobResultT> currentDoneFuture = completionService.take();
+        updateJobsStats();
+
+        JobResultT result = null;
         try {
-          Thread.sleep(JOB_POLLING_PERIOD_MS);
-        } catch (InterruptedException e) {
-          throw new BatfishException("interrupted while sleeping", e);
+          // getting the result of the job
+          result = currentDoneFuture.get();
+        } catch (InterruptedException | ExecutionException e) {
+          throw new BatfishException("Error executing job", e);
+        }
+        boolean jobResultError = handleJobResult(result, output, answerElement, failureCauses);
+        if (jobResultError) {
+          processingError = true;
         }
+
+      } catch (InterruptedException e) {
+        throw new BatfishException("interrupted while waiting for jobs to complete", e);
       }
     }
     pool.shutdown();
     if (processingError) {
-      int numJobs = jobs.size();
-      int numFailed = numJobs - failureCauses.size();
-      int numSucceeded = numJobs - numFailed;
-      if (!_logger.isActive(BatfishLogger.LEVEL_INFO)) {
-        _logger.infof("%d jobs succeeded; %d jobs failed\n", numSucceeded, numFailed);
-      }
-      if (_haltOnProcessingError) {
-        throw new CompositeBatfishException(
-            new BatfishException("Fatal exception due to failure of at least one job"),
-            failureCauses);
-      }
+      handleProcessingError(jobs, failureCauses);
     } else if (!_logger.isActive(BatfishLogger.LEVEL_INFO)) {
       _logger.info("All jobs executed successfully\n");
     }
   }
+
+  ExecutorService getExecutorService() {
+    ExecutorService pool;
+    if (!_settings.getSequential()) {
```

Comment:
Make the code more readable -- flip the order of the `if` statements so you don't need the negation.

Reply:
done